### ISSUE TITLE
MockReliableStateManager with concurrent AddOrUpdateAsync throws Object reference not set to an instance of an object

### ISSUE THREAD

**Issue by kazepreaux** [loekd/ServiceFabric.Mocks#104]
Hi,

We have been writing some benchmark tests to check the performance of our ServiceFabric application, and ran into issues when using the MockReliableStateManager.

The issue is, when running many `AddOrUpdateAsync` in parallel, we are receiving many `Object reference not set to an instance of an object.` errors.

We have been able to reproduce it without any of our code, with the test below:

```CSharp
[Test]
public async Task LoadTest()
{
    var stateManager = new MockReliableStateManager();
    var data = await stateManager.GetOrAddAsync<IReliableDictionary<Guid, string>>("data");
    var updates = new List<Task>();
    var id = Guid.NewGuid();

    for (var i = 0; i < 1000; i++)
    {
        updates.Add(Task.Run(async () =>
        {
            using (var tx = stateManager.CreateTransaction())
            {
                var newValue = DateTime.Now.ToString();
                await data.AddOrUpdateAsync(tx, id, newValue, (key, value) => newValue).ConfigureAwait(false);

                await tx.CommitAsync().ConfigureAwait(false);
            }
        }));
    }

    await Task.WhenAll(updates);
}
```

Here is the stacktrace, where sensitive info got removed with `[..]`:

```
System.AggregateException : One or more errors occurred. (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.) (Object reference not set to an instance of an object.)
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
  ----> System.NullReferenceException : Object reference not set to an instance of an object.
   at System.Threading.Tasks.Task.WaitAllCore(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks)
   at [..]
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.GetResult()
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.BeforeAndAfterTestCommand.<>c__DisplayClass1_0.<Execute>b__0()
   at NUnit.Framework.Internal.Commands.BeforeAndAfterTestCommand.RunTestMethodInThreadAbortSafeZone(TestExecutionContext context, Action action)
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
--NullReferenceException
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Wait(Int32 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, Int64 milliseconds, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.Lock`1.Acquire(T id, LockMode lockMode, TimeSpan timeout, CancellationToken cancellationToken)
   at ServiceFabric.Mocks.ReliableCollections.LockManager`2.AcquireLock(TId id, TKey key, LockMode lockMode, TimeSpan timeout, CancellationToken token)
   at ServiceFabric.Mocks.ReliableCollections.TransactedConcurrentDictionary`2.AddOrUpdateAsync(ITransaction tx, TKey key, Func`2 addValueFactory, Func`3 updateValueFactory, TimeSpan timeout, CancellationToken cancellationToken)
   at [..]
   at [..]
```

**Comment by loekd** [comment]
Thanks for reporting this issue!
Even though the problem didn't repro using mstest, I do believe I've found the issue. It should be fixed in v [4.2.3](https://www.nuget.org/packages/ServiceFabric.Mocks/4.2.3), please let me know.

**Comment by loekd** [comment]
closed due to inactivity
